src/cryptographic_primitives/hashing/merkle_tree.rs

```diff
@@ -5,66 +5,92 @@
     License MIT: https://github.com/KZen-networks/curv/blob/master/LICENSE
 */
 
-// enabled via feature since it uses rust-crypto.
-
 use std::marker::PhantomData;
 
-use crypto::sha3::Sha3;
-use merkle::{MerkleTree, Proof};
+use digest::{Digest, Output};
+use merkle_cbt::merkle_tree::{Merge, MerkleProof, MerkleTree, CBMT};
+use serde::{Deserialize, Serialize};
 
+use crate::cryptographic_primitives::hashing::DigestExt;
+use crate::cryptographic_primitives::proofs::ProofError;
 use crate::elliptic::curves::{Curve, Point};
-/*
-pub struct MT256<'a> {
-    tree: MerkleTree<GE>,
-    root: & 'a Vec<u8>,
-}
-*/
-pub struct MT256<E: Curve> {
-    tree: MerkleTree<[u8; 32]>,
-    _ph: PhantomData<E>,
+
+pub struct MT256<E: Curve, H: Digest> {
+    tree: MerkleTree<Output<H>, MergeDigest<H>>,
+    leaves: Vec<Point<E>>,
+    _ph: PhantomData<fn(E)>,
 }
 
-//impl <'a> MT256<'a>{
-impl<E: Curve> MT256<E> {
-    pub fn create_tree(vec: &[Point<E>]) -> MT256<E> {
-        let digest = Sha3::keccak256();
-        let vec_bytes = (0..vec.len())
-            .map(|i| {
-                let mut array = [0u8; 32];
-                let bytes = vec[i].to_bytes(false);
-                array.copy_from_slice(&bytes[0..32]);
-                array
-            })
-            .collect::<Vec<[u8; 32]>>();
-        let tree = MerkleTree::from_vec::<[u8; 32]>(digest, vec_bytes);
+impl<E: Curve, H: Digest + Clone> MT256<E, H> {
+    pub fn create_tree(leaves: Vec<Point<E>>) -> Self {
+        let hashes = leaves
+            .iter()
+            .map(|leaf| H::new().chain_point(leaf).finalize())
+            .collect::<Vec<_>>();
 
         MT256 {
-            tree,
+            tree: CBMT::<Output<H>, MergeDigest<H>>::build_merkle_tree(&hashes),
+            leaves,
             _ph: PhantomData,
         }
     }
 
-    pub fn gen_proof_for_ge(&self, value: &Point<E>) -> Proof<[u8; 32]> {
-        let mut array = [0u8; 32];
-        let pk_slice = value.to_bytes(false);
-        array.copy_from_slice(&pk_slice[0..32]);
-        MerkleTree::gen_proof::<[u8; 32]>(&self.tree, array).expect("not found in tree")
+    pub fn build_proof(&self, point: Point<E>) -> Option<Proof<E, H>> {
+        let index = (0u32..)
+            .zip(&self.leaves)
```

Comment:
I think `enumerate()` is more idiomatic

Reply:
although from codegen standpoint(https://godbolt.org/z/3YKPqE6e6) this is better, and there's probably no way that there are more than 2^32 leaves here (although maybe assert that in `create_tree`)

Reply:
`enumerate` yields `usize` that then needs to be converted to `u32` via cast expression `n as u32`, or conversion trait `u32::try_from(n).unwrap()`. I find zipping more pretty among all these options

Reply:
Yeah, I think there's no way number of leaves can overflow u32